### PR TITLE
Issue/m2 wp media library

### PR DESCRIPTION
**Pull request by nbradbury**
Adds the WP media library as a source for product images.

### Testing

- Open product detail and tap to add an image
- Tap "Add photos"
- Choose "WordPress media library" as the source
- Choose a few images and verify the activity title updates to show the selection count
- Press the back button and verify the discard dialog appears
- Long press an image and verify it shows fullscreen
- Rotate the device and ensure the selection is maintained
- Tap "Done" and verify the selected images appear on the product images screen
- Return to product detail, tap "Done" and verify the selected images are actually attached to the product

### Demos

Light mode and dark mode are very similar.

light

ezgif-4-b9dfbe17b75f

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by AmandaRiu — CHANGES_REQUESTED** [review]
@nbradbury This is looking really nice. Some issues I found:

- The “No images yet” text was hidden in light mode since the text was white and the background was set to `color_surface` which is white in light mode. 
- Opening an image full screen when other items are selected will lose selected items when returning to the image grid.

I started leaving code comments, but it ended up being easier to just [create a temp branch containing all my recommended changes for you to play with](https://github.com/woocommerce/woocommerce-android/compare/issue/m2-wp-media-library...amanda/m2-wp-media-library). I implemented light/dark theming based on how we are currently handling it in the main app, but @Garance91540 should definitely look it over and approve. 

I tweaked the selector dialog to use the design margins, as well as `Subtitle1` for the title since that's what's used in the rest of the app. 

Here are some screenshots from my changes:

Light | Dark
-- | --
Screenshot_1588211497|Screenshot_1588211506
Screenshot_1588211487|Screenshot_1588211476

**Comment by Garance91540** [comment]
hi @nbradbury  and @AmandaRiu , looking good on my end! 2 comments:
1. is it possible to only show the "Done" top nav CTA when there's at least one image selected?
2. I noticed that both on Woo web and the WP app there's also a search feature in the top nav. not a priority but are we thinking of including it later?

**Comment by nbradbury** [comment]
> * is it possible to only show the "Done" top nav CTA when there's at least one image selected?

@Garance91540 We can do that, but it would be inconsistent with all the other settings screens which always show "Done" regardless of changes.



> 2. I noticed that both on Woo web and the WP app there's also a search feature in the top nav. not a priority but are we thinking of including it later?

Honestly, I'd forgotten all about that feature even though I coded it for WPAndroid :) I don't think it's a priority but if as a team we want to add it, we certainly can.

**Comment by nbradbury** [comment]
> I started leaving code comments, but it ended up being easier to just [create a temp branch containing all my recommended changes for you to play with](https://github.com/woocommerce/woocommerce-android/compare/issue/m2-wp-media-library...amanda/m2-wp-media-library)

@AmandaRiu Thanks for the thorough review and style changes! I pulled your temp branch into this PR, but I kept the always dark background for `WPMediaViewerActivity` so it matches the existing `ProductImageViewerFragment` (as well as WPAndroid).

**Comment by Garance91540** [comment]
> @Garance91540 We can do that, but it would be inconsistent with all the other settings screens which always show "Done" regardless of changes.

mmm... I guess I miss this during testint. Ideally the feedback CTA is only displayed when an action is triggered. If we haven't implemented it this way for other screens we can leave it like this for now! (I may create a separate issue later to address this issue on all the screens at once)

> Honestly, I'd forgotten all about that feature even though I coded it for WPAndroid :) I don't think it's a priority but if as a team we want to add it, we certainly can.

Totally agree that's not a priority! But nice to have later probably

**Comment by nbradbury** [comment]
> I may create a separate issue later to address this issue on all the screens at once

@Garance91540 Handling this in a separate PR would definitely be preferable, as this one is already quite large.

**Comment by nbradbury** [comment]
@AmandaRiu I fixed the issue with selected images not being retained in 6dd4b01. The problem was that the fragment's `onSaveInstanceState()` wasn't being called, so the selection wasn't being saved. I resolved this by having the `WPMediaGalleryView` retain & restore its own state.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/wpmediapicker/WPMediaPickerRepository.kt [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    fun onCleanup() {
+        dispatcher.unregister(this)
+    }
+
+    /**
+     * Submits a fetch request to get a list of media for the current site
+     */
+    suspend fun fetchSiteMediaList(loadMore: Boolean = false): List<Product.Image> {
+        try {
+            loadContinuation?.cancel()
```
[the remark this answers is not in the document]
Comment (AmandaRiu) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/wpmediapicker/WPMediaPickerRepository.kt:50]: We tested this out further and it's working as expected 👍

**Review by AmandaRiu — APPROVED** [review]
Works beautifully 😍

**Comment by Garance91540** [comment]
> @Garance91540 Handling this in a separate PR would definitely be preferable, as this one is already quite large.

Oh yes, for sure!

**Pull request merged**